### DISCUSSION TITLE
.spec.postRenderers[0].kustomize.patchesStrategicMerge missing

### DISCUSSION THREAD

**Discussion by masterkain** [fluxcd/flux2#4796] [Q&A] [answered]
```
"error":"HelmRelease/kube-system/docker-proxy dry-run failed: failed to prune fields: failed add back owned items: failed to convert merged object at version helm.toolkit.fluxcd.io/v2: .spec.postRenderers[0].kustomize.patchesStrategicMerge: field not declared in schema
```

which memo did I miss for 2.3.0? thanks

**Comment by masterkain** [comment]
also this

```
"Kustomization":{"name":"flux-system","namespace":"flux-system"},"namespace":"flux-system","name":"flux-system","reconcileID":"c17ebcd3-8676-42aa-b17c-c440496d3b8f","revision":"main@sha1:ff23c740d55ac1fb030680c3758ce90da89085ad","error":"CustomResourceDefinition/helmcharts.source.toolkit.fluxcd.io dry-run failed (Invalid): CustomResourceDefinition.apiextensions.k8s.io \"helmcharts.source.toolkit.fluxcd.io\" is invalid: status.storedVersions[1]: Invalid value: \"v1\": must appear in spec.versions"}`
``

**Reply by kingdonb** [reply]
This error tells you are trying to roll back to the earlier version of Flux. Kubernetes does not support CRD downgrading

> status.storedVersions[1]: Invalid value: \"v1\": must appear in spec.versions

This is the new CRD version, Kubernetes is complaining (I think) because it is the version that CRD currently has in use (`servedVersion`) but it is not present in the version that you're trying to upgrade to. This just means what I said above; there is no safe downgrade path for CRD versions in Kubernetes. It should be safest to `flux uninstall` and re-bootstrap from the previous Flux version that you want to downgrade to, if you can't complete the upgrade now. The `flux uninstall` process is/should be safe for your workloads, as long as nothing else of critical importance lives in the `flux-system` namespace with your Flux installation.

**Comment by souleb** [comment, marked as answer]
see https://fluxcd.io/blog/2024/05/flux-v2.3.0/#breaking-changes-and-deprecations in the new release blog post.

**Reply by masterkain** [reply]
yeah I'm in the process of doing that but apparently I still have some issues, gathering details

**Comment by kingdonb** [comment]
Please take a look at the API Changes in the release notes:

> https://github.com/fluxcd/flux2/releases#:~:text=API-,changes,-HelmRelease%20v2

> `.spec.postRenderers.kustomize.patchesStrategicMerge` replaced by `.spec.postRenderers.kustomize.patches`.

You can check the previous minor, Flux 2.2.0, for the next most recent place this deprecation was mentioned, and the 2.0.0 release notes to see where this was first mentioned (I think?) - this is a deprecation that came from Kustomize upstream, `patches` can do everything that these other patch strategies can do put together. I do recommend you get in the habit of checking for breaking changes in the release notes at least every minor version upgrade, everything is covered in there. Also, the release blog.

**Reply by masterkain** [reply]
thanks, can you please check what are the apiVersion in `flux-system/gotk-sync.yaml` ?

**Reply by kingdonb** [reply]
That will depend on what versions you're upgrading/downgrading from and to. (In all Flux versions since 2.0.0, you should find that file has these API versions:

`apiVersion: source.toolkit.fluxcd.io/v1`
`apiVersion: kustomize.toolkit.fluxcd.io/v1`

These are the `v1` APIs in "Bootstrap GA", which have been marked v1 since 2.0.0-rc.1

The errors you're seeing talk about `v1` for HelmRelease and HelmRepository, I think, which both became GA (`v1`) in 2.3.0.

**Comment by masterkain** [comment]
thanks everyone for the quick help, I think i was able to upgrade also the most kustomized releases on my cluster
